vald/evm/evm.go

```diff
@@ -262,79 +264,49 @@ func (mgr Mgr) ProcessGatewayTxConfirmation(event *types.ConfirmGatewayTxStarted
 
-			if err := gatewayEvent.ValidateBasic(); err != nil {
-				mgr.logger().Debug(sdkerrors.Wrap(err, "invalid event ContractCall").Error())
-				continue
-			}
+	return err
+}
 
-			events = append(events, types.Event{
-				Chain: event.Chain,
-				TxID:  event.TxID,
-				Index: uint64(i),
-				Event: &types.Event_ContractCall{
-					ContractCall: &gatewayEvent,
-				},
-			})
-		case ContractCallWithTokenSig:
-			gatewayEvent, err := DecodeEventContractCallWithToken(txlog)
-			if err != nil {
-				mgr.logger().Debug(sdkerrors.Wrap(err, "decode event ContractCallWithToken failed").Error())
-				continue
-			}
+// ProcessGatewayTxsConfirmation votes on the correctness of an EVM chain gateway's transactions
+func (mgr Mgr) ProcessGatewayTxsConfirmation(event *types.ConfirmGatewayTxsStarted) error {
+	if !slices.Any(event.Participants, func(v sdk.ValAddress) bool { return v.Equals(mgr.validator) }) {
+		f := slices.Map(event.PollMappings, func(m types.PollMapping) voteTypes2.PollID { return m.PollID })
+		mgr.logger("poll_ids", f).Debug("ignoring gateway txs confirmation poll: not a participant")
+		return nil
+	}
 
-			if err := gatewayEvent.ValidateBasic(); err != nil {
-				mgr.logger().Debug(sdkerrors.Wrap(err, "invalid event ContractCallWithToken").Error())
-				continue
-			}
+	txIDs := slices.Map(event.PollMappings, func(poll types.PollMapping) common.Hash { return common.Hash(poll.TxID) })
+	txReceipts, err := mgr.GetTxReceiptsIfFinalized(event.Chain, txIDs, event.ConfirmationHeight)
+	if err != nil {
+		return err
+	}
 
-			events = append(events, types.Event{
-				Chain: event.Chain,
-				TxID:  event.TxID,
-				Index: uint64(i),
-				Event: &types.Event_ContractCallWithToken{
-					ContractCallWithToken: &gatewayEvent,
-				},
-			})
-		case TokenSentSig:
-			gatewayEvent, err := DecodeEventTokenSent(txlog)
-			if err != nil {
-				mgr.logger().Debug(sdkerrors.Wrap(err, "decode event TokenSent failed").Error())
-			}
+	votes := make([]sdk.Msg, len(txReceipts))
+	for i, result := range txReceipts {
+		pollID := event.PollMappings[i].PollID
+		txID := event.PollMappings[i].TxID
 
-			if err := gatewayEvent.ValidateBasic(); err != nil {
-				mgr.logger().Debug(sdkerrors.Wrap(err, "invalid event TokenSent").Error())
-				continue
-			}
+		keyvals := []interface{}{"chain", event.Chain, "poll_id", pollID.String(), "tx_id", txID.Hex()}
+		if result.Err() != nil {
+			mgr.logger(keyvals...).Infof("broadcasting empty vote due to error: %s", result.Err().Error())
+			votes[i] = voteTypes.NewVoteRequest(mgr.proxy, pollID, types.NewVoteEvents(event.Chain))
+		}
 
-			events = append(events, types.Event{
-				Chain: event.Chain,
-				TxID:  event.TxID,
-				Index: uint64(i),
-				Event: &types.Event_TokenSent{
-					TokenSent: &gatewayEvent,
-				},
-			})
-		default:
+		receipt := result.Ok()
+		if receipt == nil {
+			continue
```

Comment:
this can lead to `votes[i] == nil`. Is this going to cause issues with the broadcast?

Reply:
true, switched to slice declaration

Reply:
Should vote empty instead here.

Reply:
indeed